sc-memory/sc-memory/sc_agent.tpp

```diff
@@ -0,0 +1,580 @@
+      m_agentImplementationAddr,
+      ScType::EdgeAccessConstPosPerm,
+      ScKeynodes::nrel_inclusion);
+  if (!it5->Next())
+    SC_THROW_EXCEPTION(
+        utils::ExceptionInvalidState,
+        "Agent implementation `" << this->GetName()
+                                 << "` is not included to any abstract sc-agent. Check that agent implementation has "
+                                    "specified relation `nrel_inclusion`.");
+
+  return it5->Get(0);
+}
+
+template <class TScEvent, class TScContext>
+ScAddr ScAgentAbstract<TScEvent, TScContext>::GetEventClass() const
+{
+  if constexpr (!std::is_same<TScEvent, ScElementaryEvent>::value)
+    return TScEvent::eventClassAddr;
+
+  ScIterator5Ptr const it5 = m_memoryCtx.Iterator5(
+      GetAbstractAgent(),
+      ScType::EdgeDCommonConst,
+      ScType::EdgeDCommonConst,
+      ScType::EdgeAccessConstPosPerm,
+      ScKeynodes::nrel_primary_initiation_condition);
+  if (!it5->Next())
+    SC_THROW_EXCEPTION(
+        utils::ExceptionInvalidState,
+        "Abstract sc-agent `"
+            << this->GetName()
+            << "` has not primary initiation condition and result. Check that abstract sc-agent has specified "
+               "relation `nrel_primary_initiation_condition`.");
+
+  return m_memoryCtx.GetEdgeSource(it5->Get(2));
+}
+
+template <class TScEvent, class TScContext>
+ScAddr ScAgentAbstract<TScEvent, TScContext>::GetEventSubscriptionElement() const
+{
+  ScIterator5Ptr const it5 = m_memoryCtx.Iterator5(
+      GetAbstractAgent(),
+      ScType::EdgeDCommonConst,
+      ScType::EdgeDCommonConst,
+      ScType::EdgeAccessConstPosPerm,
+      ScKeynodes::nrel_primary_initiation_condition);
+  if (!it5->Next())
+    SC_THROW_EXCEPTION(
+        utils::ExceptionInvalidState,
+        "Abstract sc-agent `"
+            << this->GetName()
+            << "` has not primary initiation condition and result. Check that abstract sc-agent has specified "
+               "relation `nrel_primary_initiation_condition`.");
+
+  return m_memoryCtx.GetEdgeTarget(it5->Get(2));
+}
+
+template <class TScEvent, class TScContext>
+ScAddr ScAgentAbstract<TScEvent, TScContext>::GetActionClass() const
+{
+  ScIterator5Ptr const it5 = m_memoryCtx.Iterator5(
+      GetAbstractAgent(),
+      ScType::EdgeDCommonConst,
+      ScType::NodeConstClass,
+      ScType::EdgeAccessConstPosPerm,
+      ScKeynodes::nrel_sc_agent_action_class);
+  if (!it5->Next())
+    SC_THROW_EXCEPTION(
+        utils::ExceptionInvalidState,
+        "Abstract sc-agent `" << this->GetName()
+                              << "` has not action class. Check that abstract sc-agent has specified "
+                                 "relation `nrel_sc_agent_action_class`.");
+
+  return it5->Get(2);
+}
+
+// LCOV_EXCL_START
+template <class TScEvent, class TScContext>
+sc_bool ScAgentAbstract<TScEvent, TScContext>::CheckInitiationCondition(TScEvent const & event)
+{
+  return SC_TRUE;
+}
+
+// LCOV_EXCL_STOP
```

Comment:
This method implementation is never called, I haven't forgotten anything. 

I can't delete them, they may not be predefined. And if they are not overridden, I don't call them

Reply:
See GetCallback method implementation in ScAgent class

Reply:
I don't see the point in creating a test for a method that is already not called anywhere. I can't remove the implementation for this method in the base class, because I need to distinguish the implementation of this method in the child classes from some implementation in the base class. See checks that used in static_assert

Reply:
If you disagree with me, I can write tests to these methods

Reply:
Let's discuss it

Reply:
I covered this code with tests

Reply:
https://github.com/search?q=LCOV_EXCL_START&type=code